ingestion-beam/src/main/java/com/mozilla/telemetry/IpPrivacyDecoder.java

```diff
@@ -0,0 +1,88 @@
+import com.mozilla.telemetry.decoder.ipprivacy.ParseIp;
+import com.mozilla.telemetry.decoder.ipprivacy.RemoveAttributes;
+import com.mozilla.telemetry.ingestion.core.Constant.Attribute;
+import com.mozilla.telemetry.transforms.DecompressPayload;
+import com.mozilla.telemetry.transforms.NormalizeAttributes;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.Filter;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+
+/**
+ * Job for generating the IP privacy dataset that is based on the decoder job with
+ * unneeded steps taken out.
+ */
+public class IpPrivacyDecoder extends Sink {
+
+  /**
+   * Execute an Apache Beam pipeline.
+   *
+   * @param args command line arguments
+   */
+  public static void main(String[] args) {
+    run(args);
+  }
+
+  /**
+   * Execute an Apache Beam pipeline and return the {@code PipelineResult}.
+   *
+   * @param args command line arguments
+   */
+  public static PipelineResult run(String[] args) {
+    registerOptions(); // Defined in Sink.java
+    final IpPrivacyDecoderOptions.Parsed options = IpPrivacyDecoderOptions
+        .parseIpPrivacyDecoderOptions(PipelineOptionsFactory.fromArgs(args).withValidation()
+            .as(IpPrivacyDecoderOptions.class));
+    return run(options);
+  }
+
+  /**
+   * Execute an Apache Beam pipeline and return the {@code PipelineResult}.
+   */
+  public static PipelineResult run(IpPrivacyDecoderOptions.Parsed options) {
+    final Pipeline pipeline = Pipeline.create(options);
+    final List<PCollection<PubsubMessage>> errorCollections = new ArrayList<>();
+
+    // We wrap pipeline in Optional for more convenience in chaining together transforms.
+    Optional.of(pipeline) //
+        .map(p -> p //
+            .apply(options.getInputType().read(options)) //
+            .apply(ParseUri.of()).failuresTo(errorCollections) //
+            .apply("RestrictToMainPings",
+                Filter
+                    .by((message) -> "main".equals(message.getAttribute(Attribute.DOCUMENT_TYPE))))
+            .apply(ParseProxy.of()) //
+            .apply(ParseIp.of()) //
+            .apply(GeoCityLookup.of(options.getGeoCityDatabase(), options.getGeoCityFilter())) //
+            .apply(DecompressPayload.enabled(options.getDecompressInputPayloads())) //
+            .apply(ExtractClientIdAndDropPayload.of()).failuresTo(errorCollections) //
+            .apply(HashClientInfo.of(options.getClientIdHashKey(), options.getClientIpHashKey())) //
+            .apply(NormalizeAttributes.of())) //
+        .map(p -> p //
+            .apply(RemoveAttributes.of()) //
```

Comment:
Looking at https://gist.github.com/Ben-Wu/071c79c3b744589a9aff6bca6707c920#file-launch-dataflow-job I finally get what's going on here; I had made assumptions that were blocking me from properly digesting the approach.

I had assumed that we would use `outputTableRowFormat=payload`, in which case the attributes are all dropped and what we write to BQ is based on the payload. In that scenario, instead of `RemoveAttributes`, we would have a `BuildIpPrivacyPayload` that creates a `Map` of field names to values, and then serializes that to a JSON string as the payload and attaches a null attribute map.

Instead, I see now that you're specifying `outputTableRowFormat=decoded` in your configuration and relying on the values of the attributes. That's a perfectly valid way to go, though I believe we'll need to use `outputTableRowFormat=raw` instead to pass through each attribute directly as a field.

Reply:
Oops, I actually am using `--outputTableRowFormat=raw` in my testing.  I just forgot to update the gist.  Yes raw is needed to pass the attribute as a field.